spec/integration/federation/receive_federation_messages.rb

```diff
@@ -0,0 +1,216 @@
+require "spec_helper"
+require "diaspora_federation/test"
+
+def generate_xml(entity, remote_user, user)
+  DiasporaFederation::Salmon::EncryptedSlap.generate_xml(
+    remote_user.diaspora_handle,
+    OpenSSL::PKey::RSA.new(remote_user.encryption_key),
+    entity,
+    OpenSSL::PKey::RSA.new(user.encryption_key)
+  )
+end
+
+shared_examples_for "it deals correctly with a relayable" do
+  it "treats upstream receive correctly" do
+    entity = FactoryGirl.build(
+      entity_name,
+      parent_guid: local_message.guid,
+      diaspora_id: remote_user.person.diaspora_handle
+    )
+
+    expect(DiasporaFederation.callbacks).to receive(:trigger)
+                                              .with(:fetch_private_key_by_id, remote_user.person.diaspora_handle)
+                                              .once
+                                              .and_return(remote_user.encryption_key)
+    expect(DiasporaFederation.callbacks).to receive(:trigger)
+                                              .with(:fetch_private_key_by_post_guid, kind_of(String))
+                                              .and_call_original
+
+    Workers::ReceiveEncryptedSalmon.new.perform(user.id, generate_xml(entity, remote_user, user))
+    received_entity = klass.find_by(guid: entity.guid)
+    expect(received_entity).not_to be_nil
+    expect(received_entity.author.diaspora_handle).to eq(remote_user.person.diaspora_handle)
```

Comment:
I'd say it's okay if you access multiple properties. It's more critical in non test code anyway.